File: python/src/falcon_core/autotuner_interfaces/contexts/acquisition_context.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.acquisition_context import AcquisitionContext as _CAcquisitionContext
from falcon_core.physics.device_structures.connection import Connection
from falcon_core.instrument_interfaces.names.instrument_port import InstrumentPort
from falcon_core.physics.units.symbol_unit import SymbolUnit

class AcquisitionContext:
    """Python wrapper for AcquisitionContext."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new(cls, connection: Connection, instrument_type: str, units: SymbolUnit) -> AcquisitionContext:
        obj = cls(_CAcquisitionContext.new(connection._c if connection is not None else None, instrument_type, units._c if units is not None else None))
        obj._ref_connection = connection  # Keep reference alive
        obj._ref_units = units  # Keep reference alive
        return obj

    @classmethod
    def new_from_port(cls, port: InstrumentPort) -> AcquisitionContext:
        obj = cls(_CAcquisitionContext.new_from_port(port._c if port is not None else None))
        obj._ref_port = port  # Keep reference alive
        return obj

    @classmethod
    def from_json(cls, json: str) -> AcquisitionContext:
        return cls(_CAcquisitionContext.from_json(json))

    def connection(self, ) -> Connection:
        ret = self._c.connection()
        if ret is None: return None
        return Connection._from_capi(ret)

    def instrument_type(self, ) -> str:
        ret = self._c.instrument_type()
        return ret

    def units(self, ) -> SymbolUnit:
        ret = self._c.units()
        if ret is None: return None
        return SymbolUnit._from_capi(ret)

    def division_unit(self, other: SymbolUnit) -> AcquisitionContext:
        ret = self._c.division_unit(other._c if other is not None else None)
        return AcquisitionContext._from_capi(ret)

    def division(self, other: AcquisitionContext) -> AcquisitionContext:
        ret = self._c.division(other._c if other is not None else None)
        return AcquisitionContext._from_capi(ret)

    def match_connection(self, other: Connection) -> None:
        ret = self._c.match_connection(other._c if other is not None else None)
        return ret

    def match_instrument_type(self, other: str) -> None:
        ret = self._c.match_instrument_type(other)
        return ret

    def equal(self, b: AcquisitionContext) -> None:
        ret = self._c.equal(b._c if b is not None else None)
        return ret

    def not_equal(self, b: AcquisitionContext) -> None:
        ret = self._c.not_equal(b._c if b is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def __truediv__(self, other):
        """Operator overload for /"""
        if isinstance(other, (int, float)):
            return self.division(other)
        return NotImplemented

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, AcquisitionContext):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, AcquisitionContext):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/autotuner_interfaces/contexts/acquisitioncontext.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Instrument_interfaces.Names
open Falcon_core.Physics.Device_structures
open Falcon_core.Physics.Units

class c_acquisitioncontext (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.acquisitioncontext_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module AcquisitionContext = struct
  type t = c_acquisitioncontext

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.acquisitioncontext_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_acquisitioncontext ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.acquisitioncontext_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_acquisitioncontext ptr

  let make (connection : Connection.t) (instrument_type : string) (units : SymbolUnit.t) : t =
    ErrorHandling.multi_read [connection; units] (fun () ->
      let ptr = Capi_bindings.acquisitioncontext_create connection#raw (Capi_bindings.string_wrap instrument_type) units#raw in
      ErrorHandling.raise_if_error ();
      new c_acquisitioncontext ptr
    )

  let fromPort (port : InstrumentPort.t) : t =
    ErrorHandling.read port (fun () ->
      let ptr = Capi_bindings.acquisitioncontext_create_from_port port#raw in
      ErrorHandling.raise_if_error ();
      new c_acquisitioncontext ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.acquisitioncontext_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.acquisitioncontext_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.acquisitioncontext_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let connection (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.acquisitioncontext_connection handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let instrumentType (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.acquisitioncontext_instrument_type handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let units (handle : t) : SymbolUnit.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.acquisitioncontext_units handle#raw in
      ErrorHandling.raise_if_error ();
      new c_symbolunit result
    )

  let divisionUnit (handle : t) (other : SymbolUnit.t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.acquisitioncontext_division_unit handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_acquisitioncontext result
    )

  let division (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.acquisitioncontext_division handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_acquisitioncontext result
    )

  let matchConnection (handle : t) (other : Connection.t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.acquisitioncontext_match_connection handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let matchInstrumentType (handle : t) (other : string) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.acquisitioncontext_match_instrument_type handle#raw (Capi_bindings.string_wrap other) in
      ErrorHandling.raise_if_error ();
      result
    )

end